client/messaging/common-message-layout.tsx

```diff
@@ -34,55 +36,84 @@ const Text = styled.span`
   overflow: hidden;
 `
 
-export function ParsedText({ text }: { text: string }) {
-  const parsedText = useMemo(() => {
-    const matches = matchLinks(text)
+const UserMention = styled.span`
+  background-color: ${rgba(blue400, 0.48)};
+
+  &:hover {
+    background-color: ${rgba(blue400, 0.87)};
+  }
+`
+
+function* getAllMatches(text: string) {
+  yield* matchMentionsMarkup(text)
+  yield* matchLinks(text)
+}
+
+export const TextMessage = React.memo<{
+  userId: SbUserId
+  selfUserId: SbUserId
+  time: number
+  text: string
+}>(props => {
+  const { userId, selfUserId, time, text } = props
+  const [parsedText, isHighlighted] = useMemo(() => {
+    const matches = getAllMatches(text)
+    const sortedMatches = Array.from(matches).sort((a, b) => a.index - b.index)
     const elements = []
     let lastIndex = 0
+    let isHighlighted = false
 
-    for (const match of matches) {
+    for (const match of sortedMatches) {
       // Insert preceding text, if any
       if (match.index! > lastIndex) {
         elements.push(text.substring(lastIndex, match.index))
       }
 
-      elements.push(
+      if (match.type === 'mentionMarkup') {
+        const userId = Number(match.groups.userId) as SbUserId
+        if (userId === selfUserId) {
+          isHighlighted = true
+        }
+
+        elements.push(
+          match.groups.prefix,
+          <UserMention key={match.index}>
+            <ConnectedUsername userId={userId} isMention={true} />
+          </UserMention>,
+          match.groups.postfix,
+        )
+      } else if (match.type === 'link') {
         // TODO(tec27): Handle links to our own host specially, redirecting to the correct route
         // in-client instead
         // TODO(2Pac): Show a warning message about opening untrusted links
-        <a key={match.index} href={match[0]} target='_blank' rel='noopener nofollow'>
-          {match[0]}
-        </a>,
-      )
+        elements.push(
+          <a key={match.index} href={match.text} target='_blank' rel='noopener nofollow'>
+            {match.text}
+          </a>,
+        )
+      }
 
-      lastIndex = match.index! + match[0].length
+      lastIndex = match.index! + match.text.length
     }
```

Comment:
I think since these types are now effectively a discriminated union, you can add an `else` here with a call to `assertUnreachable`. (For reference, the thing you pass into `assertUnreachable` is the thing you expect TS to determine as being `never` at that point, so here it is probably `assertUnreachable(match)`)